spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala

```diff
@@ -1,65 +1,157 @@
+    val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
+    Seq(cluster, database, table)
   }
 
-  override def registerTable(tableIdentifier: Seq[String], plan: LogicalPlan): Unit = ???
+  override def registerTable(tableIdentifier: Seq[String], plan: LogicalPlan): Unit = {
+    val tableIdent = fullTableIdent(tableIdentifier)
+    cachedDataSourceTables.put(tableIdent, plan)
+  }
 
-  override def unregisterTable(tableIdentifier: Seq[String]): Unit = ???
+  override def unregisterTable(tableIdentifier: Seq[String]): Unit = {
+    val tableIdent = fullTableIdent(tableIdentifier)
+    cachedDataSourceTables.invalidate(tableIdent)
+  }
 
-  override def unregisterAllTables(): Unit = ???
+  override def unregisterAllTables(): Unit = {
+    cachedDataSourceTables.invalidateAll()
+  }
 
   override def tableExists(tableIdentifier: Seq[String]): Boolean = {
     val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
-    val schema = schemas.get(cluster)
-    val tabDef = for (ksDef <- schema.keyspaceByName.get(database); tabDef <- ksDef.tableByName.get(table)) yield tabDef
-    tabDef.isDefined
+    val tableRef = TableRef(table, database, Option(cluster))
+    val tableIdent = fullTableIdent(tableIdentifier)
+    val cached = cachedDataSourceTables.asMap().containsKey(tableIdent)
+    if (cached) {
+      true
+    } else {
+      val schema = Schema.fromCassandra(getCassandraConnector(tableRef))
+      val tabDef =
+        for (ksDef <- schema.keyspaceByName.get(database);
+             tabDef <- ksDef.tableByName.get(table)) yield tabDef
+      tabDef.nonEmpty
+    }
+  }
+
+  override def getTables(databaseName: Option[String]): Seq[(String, Boolean)] = {
+    val cluster = csc.getCluster
+    val tableNamesFromCache = getTablesFromCache(databaseName, Option(cluster)).map(_._1)
+    val tablesFromCassandra = getTablesFromCassandra(databaseName)
+    val tablesOnlyInCache =
+      tableNamesFromCache.diff(tablesFromCassandra.map(_._1)).map(name => (name, true))
+
+    tablesFromCassandra ++ tablesOnlyInCache
+  }
+
+  /** List all tables for a given database name and cluster directly from Cassandra */
+  def getTablesFromCassandra(databaseName: Option[String]): Seq[(String, Boolean)] = {
+    val cluster = csc.getCluster
+    val tableRef = TableRef("", databaseName.getOrElse(""), Option(cluster))
+    val schema = Schema.fromCassandra(getCassandraConnector(tableRef))
+    if (databaseName.nonEmpty) {
+      val ksDef = schema.keyspaceByName.get(databaseName.get)
+      if (ksDef.nonEmpty) {
+        ksDef.get.tables.map(
+          tableDef => (s"${ksDef.get.keyspaceName}.${tableDef.tableName}", false)).toSeq
+      } else {
+        Seq.empty
+      }
+    } else {
+      val tables = mutable.Set[(String, Boolean)]()
+      for (ksDef <- schema.keyspaces) {
+        tables ++= ksDef.tables.map(
+          table => (s"${ksDef.keyspaceName}.${table.tableName}", false))
+      }
+      tables.toSeq
+    }
```

Comment:
Why not simpler:
```scala
    val schema = Schema.fromCassandra(getCassandraConnector(tableRef), databaseName)
    for {
      ksDef <- schema.keyspaces.toSeq
      tableDef <- ksDef.tables
    } yield (s"${ksDef.keyspaceName}.${tableDef.tableName}", false) 
```

Schema.fromCassandra can fetch schema partially, only for the given keyspace.